manager/src/main/java/com/redhat/service/bridge/manager/workers/resources/ConnectorWorker.java

```diff
@@ -0,0 +1,176 @@
+package com.redhat.service.bridge.manager.workers.resources;
+
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.Objects;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.transaction.Transactional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.openshift.cloud.api.connector.models.Connector;
+import com.openshift.cloud.api.connector.models.ConnectorState;
+import com.openshift.cloud.api.connector.models.ConnectorStatusStatus;
+import com.redhat.service.bridge.infra.models.dto.ManagedResourceStatus;
+import com.redhat.service.bridge.manager.RhoasService;
+import com.redhat.service.bridge.manager.connectors.ConnectorsApiClient;
+import com.redhat.service.bridge.manager.dao.ConnectorsDAO;
+import com.redhat.service.bridge.manager.models.ConnectorEntity;
+import com.redhat.service.bridge.manager.models.Work;
+import com.redhat.service.bridge.rhoas.RhoasTopicAccessType;
+
+import io.quarkus.hibernate.orm.panache.PanacheRepositoryBase;
+
+@ApplicationScoped
+public class ConnectorWorker extends AbstractWorker<ConnectorEntity> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConnectorWorker.class);
+
+    @Inject
+    ConnectorsDAO connectorsDAO;
+
+    @Inject
+    RhoasService rhoasService;
+
+    @Inject
+    ConnectorsApiClient connectorsApi;
+
+    @Override
+    protected PanacheRepositoryBase<ConnectorEntity, String> getDao() {
+        return connectorsDAO;
+    }
+
+    @Override
+    public ConnectorEntity createDependencies(Work work, ConnectorEntity connectorEntity) {
+        LOGGER.info("Creating dependencies for '{}' [{}]",
+                connectorEntity.getName(),
+                connectorEntity.getId());
+
+        // This is idempotent as it gets overridden later depending on actual state
+        connectorEntity.setStatus(ManagedResourceStatus.PROVISIONING);
+        connectorEntity = persist(connectorEntity);
+
+        // Step 1 - Create Kafka Topic
+        LOGGER.debug("Creating Kafka Topic for '{}' [{}]",
+                connectorEntity.getName(),
+                connectorEntity.getId());
+        rhoasService.createTopicAndGrantAccessFor(connectorEntity.getTopicName(), RhoasTopicAccessType.PRODUCER);
+
+        // Step 2 - Create Connector
+        LOGGER.debug("Creating Managed Connector for '{}' [{}]",
+                connectorEntity.getName(),
+                connectorEntity.getId());
+        Connector connector = connectorsApi.getConnector(connectorEntity);
+        if (Objects.isNull(connector)) {
+            LOGGER.debug("Managed Connector for '{}' [{}] not found. Provisioning...",
+                    connectorEntity.getName(),
+                    connectorEntity.getId());
+            // This is an asynchronous operation so exit and wait for it's READY state to be detected on the next poll.
+            return deployConnector(connectorEntity);
```

Comment:
This is sort of a valid concern, but possibly not for your reasoning.

When a request for a `Processor` with Managed Connector is made we store the `Processor` in the database set it's status to `ACCEPTED` and store a `Connector` in the database and set its status to `PROVISION`. We then schedule some work to be performed. When the Work runs for the first time there is no contention and we end up at this call to `getConnector(..)`. If another replica (I assume you mean container/pod/server) then executes the work queue it has a different `WorkerId` and hence will not select existing work that is assigned to a different replica/container/pod/server so there should be no race-condition there.

However what happens if we have two different `Processors` both wanting to use the same Managed Connector.. assuming that's a valid use-case we could potentially have a problem there as both Workers could try to create the same Managed Connector at slightly different times. Thinking about it though.. both Workers could make the request to create a MC... one would succeed, the other would fail (as the MC API would report it already exists). When the failed Worker is re-tried it would then find the existing Connector and continue OK.

Does this answer your question?

Reply:
+1 to @manstis explanation of race condition.

We have no requirement at this point to allow Processors to share the same Managed Connector instance and this is entirely an internal implementation detail of how the Fleet Manager provisions Managed Connectors to fulfil customer requests, so again I can see ways that we can prevent race conditions here as well e.g. simply selecting a pre-deployed Managed Connector rather than trying to deploy one.

Reply:
I was thinking more about a race condition during the retry but looking better at the code it looks not possible (or easily possible) because first there is the `reconnectDroppedWorkers` that assigns a worker and then the `processWorkQueue` that assures only one worker starts.
The consistency of the system is the consequence of a set of "independent" steps but it seems correct so thanks for the clarification 👍